### PR TITLE
Refactor dimension association tables

### PR DESCRIPTION
**Pull request by daniel-thom**
This PR refactors dimension association tables in preparation for registering AEO growth rates.

- Store dimension requirements for each dataset in the project.toml.
- Cache the tables in the spark warehouse so that they don't have to be regenerated at every dataset submission.
- I also added a more generic way to register projects and datasets.

### CODE REVIEW

File: dsgrid/config/project_config.py
Comment (daniel-thom) [inline dsgrid/config/project_config.py:235]: Would it be better to define these at a level higher than dataset so that the requirements can be shared across datasets of the same category?

File: .gitmodules
Comment (lixiliu) [inline .gitmodules:4]: switch back to main when ready to merge

File: dsgrid/config/project_config.py
Comment (lixiliu) [inline dsgrid/config/project_config.py:847]: Is `existing` getting used later? Did I miss it?

File: dsgrid/config/project_config.py
Comment (lixiliu) [inline dsgrid/config/project_config.py:922]: one record is singular the other is plural?

File: dsgrid/registry/project_registry_manager.py
Comment (lixiliu) [inline dsgrid/registry/project_registry_manager.py:384]: what if dt_str contains "space", "dash", or "underscore"?

File: dsgrid/tests/registration_models.py
Comment (lixiliu) [inline dsgrid/tests/registration_models.py:9]: Why is this in the `tests` folder?

**Review by lixiliu — COMMENTED** [review]
Is there a companion PR in SS for the paradigm shift in the dimension association creation?

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:847]: line 851

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:922]: This function replaces one supplemental dimension record ID with its corresponding base dimension record IDs.

File: dsgrid/registry/project_registry_manager.py
Comment (daniel-thom) [reply dsgrid/registry/project_registry_manager.py:384]: `dt_str` will be a `DimensionType` value, and so it shouldn't have a space or dash. I suppose a future dimension type could have an underscore. I don't think I follow your concern. Can you elaborate?

File: dsgrid/tests/registration_models.py
Comment (daniel-thom) [reply dsgrid/tests/registration_models.py:9]: I expect that this will only be used in test code or by dsgrid developers that are building registries for test purposes.

**Comment by daniel-thom** [comment]
> Is there a companion PR in SS for the paradigm shift in the dimension association creation?

I have a local branch but want to wait until we merge this + the dsgrid-test-data PR. There is one pending design question.

**Comment by mooneyme** [comment]
I added this to a team's chat with dan, but also posting here:

....

I like where this is going!

My main concern here is that with this refactor is centered around messaging of the required dimensions per dataset. At the moment, we are using supplemental dimensions to define required dimensions for a dataset, e.g.:

```
[datasets.required_dimensions]
[datasets.required_dimensions.single_dimensional]
subsector = [ "residential_subsectors",]
metric = [ "residential_end_uses",]
sector = [ "res",]
```

When we declare the dataset dimension requirements in this way, it can be misleading and it may seem like we literally are looking for one subsector called `"residential_subsectors"` from the resstock data. In reality, we want the dataset to have all the residential subsectors that the project's base dimension defines (or at least provide a mapping to the base dimension).

I want it to be clear that we are expecting a subset from project’s base dimension. How can we make that more clear and less confusing for a resstock modeler? Maybe we can have a specific naming convention for these kinds of supplemental dimensions? Maybe `"supplemental.residential_subsectors"` or something like that.

File: dsgrid/tests/registration_models.py
Comment (mooneyme) [reply dsgrid/tests/registration_models.py:9]: Can we add a comment at the top of expected use case for this script?

File: dsgrid/config/project_config.py
Comment (daniel-thom) [inline dsgrid/config/project_config.py:235]: Something that might be confusing about this implementation is that we use a supplemental dimension record ID to uniquely identify supplemental dimensions while in other places we use a dimension query name. There is a built-in assumption that there is only one record ID for these type of supplemental dimensions. It could be particularly confusing because the record ID is likely to be the same as the dimension query name.

An alternative solution is to have one instance of RequiredDimensionsModel for base dimension records and a second for supplemental dimension query names.

**Comment by daniel-thom** [comment]
> I added this to a team's chat with dan, but also posting here:
> 
> ....
> 
> I like where this is going!
> 
> My main concern here is that with this refactor is centered around messaging of the required dimensions per dataset. At the moment, we are using supplemental dimensions to define required dimensions for a dataset, e.g.:
> 
> ```
> [datasets.required_dimensions]
> [datasets.required_dimensions.single_dimensional]
> subsector = [ "residential_subsectors",]
> metric = [ "residential_end_uses",]
> sector = [ "res",]
> ```
> 
> When we declare the dataset dimension requirements in this way, it can be misleading and it may seem like we literally are looking for one subsector called `"residential_subsectors"` from the resstock data. In reality, we want the dataset to have all the residential subsectors that the project's base dimension defines (or at least provide a mapping to the base dimension).
> 
> I want it to be clear that we are expecting a subset from project’s base dimension. How can we make that more clear and less confusing for a resstock modeler? Maybe we can have a specific naming convention for these kinds of supplemental dimensions? Maybe `"supplemental.residential_subsectors"` or something like that.

I added a note about an alternative solution to the companion dsgrid PR. If we want to get more specific about base vs supplemental, I think we should do it in the data structure definition rather than inventing text formats that need to be decoded.

File: dsgrid/config/dimension_association_manager.py
Comment (elainethale) [inline dsgrid/config/dimension_association_manager.py:46]: Is it correct that we used to have a project-wide association table, but now we only have dataset-specific association tables? I guess it is a simple matter to concatenate the dataset-specific association tables with the same pivoted dimension to get a project-wide association table?

File: dsgrid/registry/registry_manager_base.py
Comment (elainethale) [inline dsgrid/registry/registry_manager_base.py:231]: I thought we got rid of VersionInfo? Do we just use it for a moment to calculate the new version string?

File: dsgrid/api/app.py
Comment (daniel-thom) [reply dsgrid/api/app.py:68]: The default Hive metastore uses Apache Derby which only allows one user. Also, our standalone cluster on Eagle currently only allows one user (that could be improved if we are more particular about allocating cores). So, if the main process on the API server creates a session, receives a request to run a query, starts a child process which will also try to create a session, it will be deadlocked.

File: dsgrid/config/dimension_association_manager.py
Comment (daniel-thom) [reply dsgrid/config/dimension_association_manager.py:46]: We used to build a project-wide table but then filter it by a data source before evaluating it, _I think_. I don't think we have a need for a project-wide table, but please correct me if I'm wrong.

File: dsgrid/registry/registry_manager_base.py
Comment (daniel-thom) [reply dsgrid/registry/registry_manager_base.py:231]: We no longer use it in the Pydandic serialization/deserialization, but we still use it in the code.

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:158]: any combination of base or supplemental dimension record IDs

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:235]: Ah, yes, query names do not exist yet, so this idea isn't great. I started off with record IDs because that is what is used for mapping between base and supplemental. As it happens, these supplemental record IDs and query names will be either be identical or very close. We are definitely setting up our users for some confusion.

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:237]: No, any combination of base or supplemental dimension _record_ ID.

Comment (daniel-thom) [reply dsgrid/config/project_config.py:237]: Here is a specific example:

In this paradigm StandardScenarios would define this supplemental dimension:
```
name = "residential-subsectors"
display_name = "Residential Subsectors"
dimension_query_name = "residential_subsectors"
type = "subsector"
id = "residential-subsectors__656c0d22-c47e-4f95-a96e-1e31c17d8d16"
```
It would have one record:
```
id,name
residential_subsectors,Residential Subsectors
```
It would have these mapping records:
```
from_id,to_id
mobile_home,residential_subsectors
multifamily_2_to_4_units,residential_subsectors
multifamily_5_plus_units,residential_subsectors
single_family_attached,residential_subsectors
single_family_detached,residential_subsectors
```
`name`, `id`, `dimension_query_name`, `display_name`, and `record_id` are all very similar.

Comment (elainethale) [reply dsgrid/config/project_config.py:237]: Oh--that is weirder than I thought. I was thinking we could list the supplemental name, id, or query_name so that any base dimension record id that gets mapped to that supplemental dimension would be included.

Being able to reference any supplemental dimension record id also seems to set us up for name conflicts/ambiguities.

Per comments above, maybe we should use different data models for specifying requirements via (a) base dimension record ids versus (b) supplemental dimensions. The latter needs to identify the specific dimension (by name, id, or query name) and then could potentially also let users specify specific record ids.

File: dsgrid/config/dimension_association_manager.py
Comment (elainethale) [reply dsgrid/config/dimension_association_manager.py:46]: I don't think we immediately need a project-wide table, and maybe it's never a good idea to construct the whole thing

File: dsgrid/config/project_config.py
Comment (daniel-thom) [reply dsgrid/config/project_config.py:235]: The default option is here: define one set of requirements per dataset. The downside is that two datasets cannot share the same requirements.

Other options:
1. Define requirements using a field that exists in each dataset config. `data_source`, `sector_description`, some new term. The field must be an enum.
2. Allow the project coordinator to define this free-form. The list of requirements would be at the top level of the project config. Each set of requirements would have a name. Let's suppose one is called `residential`.  Each dataset in `datasets` would have an optional field called `dimension_requirements`. ResStock and NEMS Residential would have `residential.` If the field is not provided then all project dimension records must be present.

Comment (elainethale) [reply dsgrid/config/project_config.py:254]: We discussed and decided to leave under dataset for now, in part because some cases that look like reuse might actually involve some tweaks.

Comment (elainethale) [reply dsgrid/config/project_config.py:254]: The revision distinguishes between base and supplemental dimensions so the latter can be clearly referenced (thus avoiding name conflict issues).

**Review by elainethale — APPROVED** [review]

File: dsgrid/registry/project_registry_manager.py
Comment (lixiliu) [reply dsgrid/registry/project_registry_manager.py:384]: Not super important, just wondering if we also need to do dt_str.replace() according to the "space", "dash", "underscore" convention in the name.

Comment (daniel-thom) [reply dsgrid/registry/project_registry_manager.py:384]: The only issue would be the possibility of a future dimension value containing an underscore. I don't think we have any rules about underscores in dimension/dimension_mapping record value IDs or names. Right?

**Review by mooneyme — APPROVED** [review]

**Pull request merged**
